File: interview questions/branchSums.py
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def branchSums(root):
    sums_list = []
    calculateSums(root, sums_list, 0)
    return sums_list


def calculateSums(node, list, sum):
    if node is None:
        return

    sum += node.value
    if node.left is None and node.right is None:
        list.append(sum)
        return

    calculateSums(node.left, list, sum)
    calculateSums(node.right, list, sum)

File: interview questions/branchSums.js
class BinaryTree {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

function branchSums(root) {
  let results = [];
  calculateSum(root, results, 0);
  return results;
}

function calculateSum(node, results, sum) {
  if (!node) return;

  sum += node.value;
  if (!node.left && !node.right) {
    results.push(sum);
    return;
  }

  calculateSum(node.left, results, sum);
  calculateSum(node.right, results, sum);
}
